intermediate_source/TP_tutorial.rst

```diff
@@ -0,0 +1,282 @@
+
+.. code-block:: python
+    import torch.nn.functional as F
+    from torch.distributed.tensor.parallel import loss_parallel
+
+    pred = model(input_ids)
+    with loss_parallel():
+        # assuming pred and labels are of the shape [batch, seq, vocab]
+        loss = F.cross_entropy(pred.flatten(0, 1), labels.flatten(0, 1))
+        loss.backward()
+
+
+How to combine Tensor Parallel with Fully Sharded Data Parallel together
+------------
+
+
+Now that we have shown how to apply Tensor/Sequence Parallel to the model, let us also take a look at how Tensor Parallel and Fully Sharded Data Parallel could work together. Since Tensor Parallelism incurs communications that’s blocking the computation, we want to make sure it runs within a fast communication channel (i.e. NVLink). In practice, we usually apply Tensor Parallel within each host, and apply Fully Sharded Data Parallel across the hosts.
+
+.. figure:: /_static/img/distributed/fsdp_tp.png
+   :width: 100%
+   :align: center
+   :alt: fsdp + tp
+   Figure 3. FSDP and TP work on separate device dimensions, FSDP communication happens inter-host and TP communication happens intra-host.
+
+This 2-D parallelism pattern can be easily expressed via a 2-D DeviceMesh, and we just need pass each “sub” DeviceMesh to each individual parallelism APIs:
+
+.. code-block:: python
+    from torch.distributed.device_mesh import init_device_mesh
+    from torch.distributed.tensor.parallel import ColwiseParallel, RowwiseParallel, parallelize_module
+    from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+
+    # i.e. 2-D mesh is [dp, tp], training on 64 GPUs that performs 8 way DP and 8 way TP
+    mesh_2d = init_device_mesh("cuda", (8, 8))
+    tp_mesh = mesh_2d["tp"] # a submesh that connects intra-host devices
+    dp_mesh = mesh_2d["dp"] # a submesh that connects inter-host devices
+
+    model = Model(...)
+
+    # apply tensor parallel
+    model_tp = parallelize_module(model, tp_mesh, tp_plan)
+    model_2d = FSDP(model_tp, device_mesh=dp_mesh, use_orig_params=True, ...)
+
+
+This would allow us to easily apply Tensor Parallel within each host (intra-host) and apply FSDP across hosts (inter-hosts). The Tensor(Model) Parallelism and Data Parallelism techniques combined together provides the ability to continue increasing model size and efficient training using a large number of GPUs.
+
+We can see that the PyTorch Tensor Parallel makes 0-code changes to the Llama model, allowing fast enablement for prototyping and productionizations. To see the complete e2e code example about how to apply Tensor/Sequence Parallelism to the Llama model, please refer to the examples repository.
```

Comment:
Please add a Conclusion. For example:

This tutorial demonstrates how to train a large Transformer-like model across hundreds to thousands of GPUs using Tensor Parallel in combination with Fully Sharded Data Parallel. It explains the working of Tensor Parallel, which is an efficient model parallelism technique to train large scale Transformer models. 

To learn more about Tensor Parallel, see:

* Link 1
* Link 2
* Link 3